crates/cairo-lang-protostar/src/casm_generator/mod.rs

```diff
@@ -0,0 +1,287 @@
+use cairo_vm::vm::errors::vm_errors::VirtualMachineError;
+use itertools::chain;
+use num_bigint::BigUint;
+use num_traits::{Num, Signed};
+use num_integer::Integer;
+use serde::{Deserialize, Serialize};
+use num_bigint::BigInt;
+use thiserror::Error;   
+use smol_str::SmolStr;
+
+
+#[derive(Debug, Error)]
+pub enum GeneratorError {
+    #[error("Failed calculating gas usage, it is likely a call for `get_gas` is missing.")]
+    FailedGasCalculation,
+    #[error("Function with suffix `{suffix}` to run not found.")]
+    MissingFunction { suffix: String },
+    #[error("Function expects arguments of size {expected} and received {actual} instead.")]
+    ArgumentsSizeMismatch { expected: usize, actual: usize },
+    #[error(transparent)]
+    ProgramRegistryError(#[from] Box<ProgramRegistryError>),
+    #[error(transparent)]
+    SierraCompilationError(#[from] CompilationError),
+    #[error(transparent)]
+    ApChangeError(#[from] ApChangeError),
+    #[error(transparent)]
+    VirtualMachineError(#[from] Box<VirtualMachineError>),
+}
+
+
+#[derive(Default, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TestEntrypoint {
+    pub offset: usize,
+    pub name: String
+}
+
+#[derive(Default, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct ProtostarCasm {
+    #[serde(serialize_with = "serialize_big_uint", deserialize_with = "deserialize_big_uint")]
+    pub prime: BigUint,
+    pub bytecode: Vec<BigIntAsHex>,
+    pub hints: Vec<(usize, Vec<String>)>,
+    pub test_entry_points: Vec<TestEntrypoint>,
+}
+
+pub struct SierraCasmGenerator {
+    /// The sierra program.
+    sierra_program: cairo_lang_sierra::program::Program,
+    /// Program registry for the Sierra program.
+    sierra_program_registry: ProgramRegistry<CoreType, CoreLibfunc>,
+    /// The casm program matching the Sierra code.
+    casm_program: CairoProgram,
+}
+
+impl SierraCasmGenerator {
+    pub fn new(
+        sierra_program: cairo_lang_sierra::program::Program,
+        calc_gas: bool,
+    ) -> Result<Self, GeneratorError> {
+        let metadata = create_metadata(&sierra_program, calc_gas)?;
+        let sierra_program_registry =
+            ProgramRegistry::<CoreType, CoreLibfunc>::new(&sierra_program)?;
+        let casm_program =
+            cairo_lang_sierra_to_casm::compiler::compile(&sierra_program, &metadata, calc_gas)?;
+        Ok(Self { sierra_program, sierra_program_registry, casm_program })
+    }
+    
+    pub fn collect_tests(&self) -> Vec<&SmolStr> {
+        self.sierra_program
+        .funcs
+        .iter()
+        .filter(|f|
+            if let Some(name) = &f.id.debug_name { name.contains("test_") } else { false }
+        )
+        .map(|f| 
+            match &f.id.debug_name {
+                Some(name) => name,
+                _ => panic!("Expected name")
+            } 
+        )
+        .collect::<Vec<&SmolStr>>()
```

Comment:
```suggestion
        .collect()
```

compiler can infer this